dali/operators/reader/loader/fits_loader_gpu.cc

```diff
@@ -0,0 +1,104 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <dirent.h>
+#include <errno.h>
+#include <fitsio.h>
+#include <cstdlib>
+#include <memory>
+
+#include "dali/core/common.h"
+#include "dali/operators/reader/loader/fits_loader_gpu.h"
+#include "dali/operators/reader/loader/utils.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/util/file.h"
+#include "dali/util/fits.h"
+
+namespace dali {
+
+void FitsLoaderGPU::ReadSample(FitsFileWrapperGPU& target) {
+  auto filename = files_[current_index_++];
+  int status = 0, num_hdus = 0;
+
+  // handle wrap-around
+  MoveToNextShard(current_index_);
+
+  // metadata info
+  DALIMeta meta;
+  // meta.SetSourceInfo(filename); // it adds ./before a filename for some reason
+  meta.SetSkipSample(false);
+
+
+  auto path = filesystem::join_path(file_root_, filename);
+  auto current_file = fits::FitsHandle::OpenFile(path.c_str(), READONLY);
+  fits::FITS_CALL(fits_get_num_hdus(current_file, &num_hdus, &status));
+
+  // resize ouput vector according to the number of HDUs
+  target.data.resize(hdu_indices_.size());
+  target.header.resize(hdu_indices_.size());
+  target.tile_offset.resize(hdu_indices_.size());
+  target.tile_size.resize(hdu_indices_.size());
+
+  for (size_t output_idx = 0; output_idx < hdu_indices_.size(); output_idx++) {
+    // move to appropiate hdu
+    fits::FITS_CALL(fits_movabs_hdu(current_file, hdu_indices_[output_idx], NULL, &status));
+
+    // read the header
+    fits::HeaderData header;
+    try {
+      fits::ParseHeader(header, current_file);
+      target.header[output_idx] = header;
+    } catch (const std::runtime_error& e) {
+      DALI_FAIL(e.what() + ". File: " + filename);
+    }
+
+    int anynul = 0, nulval = 0;
+    Index nelem = header.size();
+
+    // reset, resize specific output in target
+    if (target.data[output_idx].shares_data()) {
+      target.data[output_idx].Reset();
+      // target.tile_size[output_idx].reset
+    }
+
+    if (header.compressed) {
+      vector<uint8_t> raw_data;
+      dali::TensorShape<-1> shape;
+
+      fits::extract_undecoded_data(current_file, raw_data, target.tile_offset[output_idx],
+                                   target.tile_size[output_idx], header.rows, &status);
+
+      shape.shape.push_back(raw_data.size());
+      target.data[output_idx].Resize(shape, DALI_UINT8);
+      memcpy(static_cast<uint8_t*>(target.data[output_idx].raw_mutable_data()), raw_data.data(),
+             raw_data.size());
+    } else {
+      target.data[output_idx].Resize(header.shape, header.type());
+
+      // copy the image to host memory
+      fits::FITS_CALL(fits_read_img(
+          current_file, header.datatype_code, 1, nelem, &nulval,
+          static_cast<uint8_t*>(target.data[output_idx].raw_mutable_data()), &anynul, &status));
+    }
```

Comment:
Maybe let's try to extract this as a helper function and unify the implementation of parsing the header, etc with the LoaderCPU.

Reply:
Done!